### PR TITLE
NCBI Assembly Reader + Parser

### PR DESCRIPTION
**Pull request by singaltanmay**
AssemblyReportReader takes an InputStreamReader with an NCBI Assembly Report file and can parse into an AssemblyEntity object containing a List of ChromosomeEntity objects. I have also added a static assembly report at src/test/resources/GCF_000003055.4_Bos_taurus_UMD_3.1_assembly_report.txt to run tests on.

### CODE REVIEW

File: src/main/java/com/ebivariation/contigalias/dus/AssemblyReportReader.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+    private AssemblyEntity assemblyEntity;
+
+    private boolean reportParsed = false;
+
+    public AssemblyReportReader(InputStreamReader inputStreamReader) {
+        reader = new BufferedReader(inputStreamReader);
+    }
+
+    private void parseReport() throws IOException {
```
Comment (jmmut) [inline src/main/java/com/ebivariation/contigalias/dus/AssemblyReportReader.java:24]: our code style prioritises higher level methods first, and the inner methods later. An example:

```
A() {
  B();
  D();
}
B() {
  C();
}
C() { }
D() { }
```
so please move getAssemblyEntity before parseReport.

File: src/main/java/com/ebivariation/contigalias/dus/AssemblyReportReader.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        if (chromosomes == null) {
+            chromosomes = new LinkedList<>();
+            assemblyEntity.setChromosomes(chromosomes);
+        }
+        chromosomes.add(chromosomeEntity);
+
+    }
+
+    public long getLineCount() {
+        return reader.lines().count();
```
Comment (jmmut) [inline src/main/java/com/ebivariation/contigalias/dus/AssemblyReportReader.java:133]: I think the `lines()` method consumes the reader, so if you have an AssemblyReportReader you can either parse the asm report *or* count its lines, and this constraint is not reflected in the class interface.

Comment (singaltanmay) [reply src/main/java/com/ebivariation/contigalias/dus/AssemblyReportReader.java:110]: How can I fix this? Currently this is only used to check if reader has loaded the data. I can just modify the test to use reader.ready() and remove this method entirely.

**Comment by singaltanmay** [comment]
> Can you download the GCA instead of the GCF? the contents should be almost identical, but that shows that we prefer INSDC.

Done

File: src/main/java/com/ebivariation/contigalias/dus/AssemblyReportReader.java (hunk 2 above)
Comment (jmmut) [reply src/main/java/com/ebivariation/contigalias/dus/AssemblyReportReader.java:149]: yes, that sounds good. use ready and remove the method.

File: src/main/java/com/ebivariation/contigalias/entities/AssemblyEntity.java [hunk 3]
```diff
@@ -1,3 +1,19 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 package com.ebivariation.contigalias.entities;
 
 import javax.persistence.*;
```
Comment (jmmut) [inline src/main/java/com/ebivariation/contigalias/entities/AssemblyEntity.java:19]: also, expand these imports (and any other we have missed).

**Pull request merged**